src/ui/public/saved_object_table/item_table_state.js

```diff
@@ -0,0 +1,38 @@
+import { SortOrder } from 'ui_framework/components/table/sort_order';
+
+export class ItemTableState {
```

Comment:
What @cjcenizal is essentially proposing is to turn the `ItemSelector` from a collection of actions into something very similar to a flux store that handles the selection state. So we're getting into the "pattern for state management" discussion again. I would consider "single state" and separate actions to be the safer, more future-proof pattern (by virtue of being just a special case that can be represented in most state management implementations I've seen). In the context view app I've implemented it by currying the action and partially applying the state in the "container component": https://github.com/elastic/kibana/blob/master/src/core_plugins/kibana/public/context/app.js#L58

Regarding the ids I would recommend to have one collection of items that contain the actual data (e.g. the `items` array in this case) and some collections from which particular views on those items can be derived (such as the selected items or the items on the current page) via simple `[].filter()` operations (your `calculateItemsOnPage` already does this in a way).

So the approach I'm proposing comes down to something like

1. think about all the state present in this part of the application
2. arrange it into a state object composed of various sub-state objects (`.items`, `.selectedItems`, `pageSize`, `pageOffset`, etc)
3. remove duplications by only keeping what's needed to derive everything else (for example `.items` with the full objects and `.selectedItems` as an array of ids)
4. think about all the ways that this state needs to be mutated
5. collect these mutations into actions grouped by concern

I'm reasonably confident that this results in code that does not depend on a particular state management library but can be refactored into one without undue effort.